src/sentry/static/sentry/app/views/organizationDashboard/discoveryQuery.jsx

```diff
@@ -0,0 +1,133 @@
+import SentryTypes from 'app/sentryTypes';
+import createQueryBuilder from 'app/views/organizationDiscover/queryBuilder';
+import withGlobalSelection from 'app/utils/withGlobalSelection';
+import withOrganization from 'app/utils/withOrganization';
+
+class DiscoverQuery extends React.Component {
+  static propTypes = {
+    compareToPeriod: PropTypes.shape({
+      statsPeriodStart: PropTypes.string,
+      statsPeriodEnd: PropTypes.string,
+    }),
+    includePreviousPeriod: PropTypes.bool,
+    organization: SentryTypes.Organization,
+    selection: SentryTypes.GlobalSelection,
+    queries: PropTypes.arrayOf(SentryTypes.DiscoverQuery),
+  };
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      results: null,
+      systemResults: null,
+    };
+
+    // Query builders based on `queries`
+    this.queryBuilders = [];
+
+    // Query builders for system generated queries
+    // e.g. when comparing to equivalent query but with different timeframe
+    this.systemQueryBuilders = [];
+
+    this.createQueryBuilders();
+  }
+
+  componentDidMount() {
+    this.fetchData();
+  }
+
+  componentDidUpdate(prevProps) {
+    if (prevProps === this.props) {
+      return;
+    }
+
+    this.fetchData();
+  }
+
+  createQueryBuilders() {
+    const {compareToPeriod, organization, queries} = this.props;
+    queries.forEach(query => {
+      this.queryBuilders.push(createQueryBuilder(this.getQuery(query), organization));
+
+      // Make sure this.queryBuilders.length === this.systemQueryBuilders.length
+      this.systemQueryBuilders.push(
+        compareToPeriod
+          ? createQueryBuilder(this.getQuery(query, compareToPeriod), organization)
+          : null
+      );
+    });
+  }
+
+  getQuery(query, compareToPeriod) {
+    const {includePreviousPeriod} = this.props;
+    const {datetime, ...selection} = this.props.selection;
+    let period;
+
+    if (!compareToPeriod) {
+      const {start, end, statsPeriod} = getPeriod(datetime, {
+        shouldDoublePeriod: includePreviousPeriod,
+      });
+      period = {start, end, range: statsPeriod};
+    }
+
+    return {
+      ...query,
+      ...selection,
+      ...period,
+      ...compareToPeriod,
+    };
+  }
+
+  resetQueries() {
+    const {compareToPeriod, queries} = this.props;
+    this.queryBuilders.forEach((builder, i) => {
+      const query = queries[i];
+      builder.reset(this.getQuery(query));
+
+      if (this.systemQueryBuilders[i]) {
+        this.systemQueryBuilders[i].reset(this.getQuery(query, compareToPeriod));
+      }
+    });
+  }
+
+  async fetchData() {
+    // Reset query builder
+    this.resetQueries();
+
+    // Fetch
+    const promises = this.queryBuilders.map(builder => builder.fetch());
+    const systemPromises = this.systemQueryBuilders.map(
+      builder => (builder ? builder.fetch() : null)
+    );
+    let results = await Promise.all(promises);
+    let systemResults = await Promise.all(systemPromises);
```

Comment:
I think I don't quite understand what results vs systemResults is supposed to be here

Reply:
It's currently unused, but the idea is that it's a generated query based on the user's query. Example being we want the same query but with a different time period (from 15 days ago to 8 days ago)